# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [3]:
#欠損値の確認
train_df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [4]:
#最頻値で埋める
train_df["Age"] = train_df["Age"].fillna(train_df["Age"].mode()[0])
train_df["Cabin"] = train_df["Cabin"].fillna(train_df["Cabin"].mode()[0])
train_df["Embarked"] = train_df["Embarked"].fillna(train_df["Embarked"].mode()[0])
#再度確認
train_df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [5]:
#性別を数字にする
sex_map = {'male': 0, 'female': 1}
train_df['Sex'] = train_df['Sex'].map(sex_map)
train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,B96 B98,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,B96 B98,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,B96 B98,S
5,6,0,3,"Moran, Mr. James",0,24.0,0,0,330877,8.4583,B96 B98,Q
6,7,0,1,"McCarthy, Mr. Timothy J",0,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",0,2.0,3,1,349909,21.0750,B96 B98,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",1,27.0,0,2,347742,11.1333,B96 B98,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",1,14.0,1,0,237736,30.0708,B96 B98,C


In [6]:
# Embarked列をOne-Hotエンコーディングします
#embarked_dummies = pd.get_dummies(train_df['Embarked'], prefix='Embarked')

# 元のデータ表に、新しく作成した3つの列を連結します
#train_df = pd.concat([train_df, embarked_dummies], axis=1)

# 不要になった元のEmbarked列は削除します
#train_df = train_df.drop('Embarked', axis=1)

# 変換後のデータを確認します
#train_df.head(10)

In [7]:
# 数値データに関する基本的な統計量を表示します
print(train_df.describe())

       PassengerId    Survived      Pclass         Sex         Age  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642    0.352413   28.566970   
std     257.353842    0.486592    0.836071    0.477990   13.199572   
min       1.000000    0.000000    1.000000    0.000000    0.420000   
25%     223.500000    0.000000    2.000000    0.000000   22.000000   
50%     446.000000    0.000000    3.000000    0.000000   24.000000   
75%     668.500000    1.000000    3.000000    1.000000   35.000000   
max     891.000000    1.000000    3.000000    1.000000   80.000000   

            SibSp       Parch        Fare  
count  891.000000  891.000000  891.000000  
mean     0.523008    0.381594   32.204208  
std      1.102743    0.806057   49.693429  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    7.910400  
50%      0.000000    0.000000   14.454200  
75%      1.000000    0.000000   31.000000  
max      8.000000

### データの規模（count）
全体のデータ数は 891行 あり、それぞれの項目（PassengerId、Survived、Pclassなど）が揃っていることが分かります。
### 生存率（Survived）
Survived 列の平均値（mean）が 約0.384（38.4%） です。これは、訓練データ全体の中で生存した人の割合が約38%であったことを示しています。
### 乗客の階級（Pclass）
Pclass（チケットクラス）の最小値が1、最大値が3、平均値が 約2.3 です。全体の分布として、1等客室よりも3等客室（下位クラス）の乗客の割合が多い傾向が読み取れます。
### 年齢の傾向（Age）
Age の平均は約28.6歳、中央値（50%）は24歳、最大値は80歳、最小値は約0.5歳（生後数ヶ月の乳児）となっています。幅広い年齢層の乗客が乗船していたことがわかります。
### 家族・同伴者の状況（SibSp / Parch）
SibSp（兄弟・配偶者の数）や Parch（両親・子供の数）の平均はいずれも0.3〜0.5付近ですが、最大値を見ると SibSp は8、Parch は6となっています。大部分の人は一人または少人数で乗船している一方で、大家族で乗っている人も一部に存在することがわかります。
### 運賃のばらつき（Fare）
Fare（運賃）の平均は約32.2ドルですが、中央値（50%）は14.5ドル、そして最大値は 512.3ドル に達しています。大部分の人は比較的安価な切符で乗っているのに対し、一部の非常に高額な切符を買った富裕層が存在するため、データが大きく右に歪んでいる（一部の超高額な外れ値がある）ことが読み取れます。

In [8]:
serving_df.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [9]:
#最頻値で埋める
serving_df["Age"] = serving_df["Age"].fillna(serving_df["Age"].mode()[0])
serving_df["Fare"] = serving_df["Fare"].fillna(serving_df["Fare"].mode()[0])
serving_df["Cabin"] = serving_df["Cabin"].fillna(serving_df["Cabin"].mode()[0])
#再度確認
serving_df.isnull().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [10]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def get_title(name):
        title = name.split(",")[1].split(".")[0].strip()
        rare_titles = ["Lady","Countess","Capt","Col","Don","Dr",
                       "Major","Rev","Sir","Jonkheer","Dona"]
        if title in rare_titles:
            return "Rare"
        mapping = {"Mlle":"Miss","Ms":"Miss","Mme":"Mrs"}
        return mapping.get(title, title)
    
    df["Title"] = df["Name"].apply(get_title)
    
    # Sex と Pclass を文字列として結合してから Sex を数値にマッピングする
    df["Sex_Pclass"] = df["Sex"].astype(str) + "_" + df["Pclass"].astype(str)
    
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    
    return df

# データの読み込み直後（欠損値埋めの後など）に、そのまま preprocess を呼ぶ
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title,Sex_Pclass,Ticket_number
0,1,0,3,Braund Mr Owen Harris,0,22.0,1,0,A/5 21171,7.2500,B96 B98,S,Mr,0_3,21171
1,2,1,1,Cumings Mrs John Bradley Florence Briggs Thayer,1,38.0,1,0,PC 17599,71.2833,C85,C,Mrs,1_1,17599
2,3,1,3,Heikkinen Miss Laina,1,26.0,0,0,STON/O2. 3101282,7.9250,B96 B98,S,Miss,1_3,3101282
3,4,1,1,Futrelle Mrs Jacques Heath Lily May Peel,1,35.0,1,0,113803,53.1000,C123,S,Mrs,1_1,113803
4,5,0,3,Allen Mr William Henry,0,35.0,0,0,373450,8.0500,B96 B98,S,Mr,0_3,373450


Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [11]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")
#input_features.remove("Embarked")     
input_features.remove("Cabin")  
input_features.remove("Sex")  
input_features.remove("Pclass")  
input_features.remove("SibSp")  
input_features.remove("Parch")  
input_features.remove("Fare") 
input_features.remove("Name")

print(f"Input features: {input_features}")

Input features: ['Age', 'Embarked', 'Title', 'Sex_Pclass', 'Ticket_number']


# Convert Pandas dataset to TensorFlow Dataset

In [12]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [13]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=2, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.compile(metrics=["accuracy"])
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

Use 4 thread(s) for training
Use /tmp/tmpkzgg2r0p as temporary training directory
Reading training dataset...
Training tensor examples:
Features: {'PassengerId': <tf.Tensor 'data_7:0' shape=(None,) dtype=int64>, 'Pclass': <tf.Tensor 'data_8:0' shape=(None,) dtype=int64>, 'Name': tf.RaggedTensor(values=Tensor("data_4:0", shape=(None,), dtype=string), row_splits=Tensor("data_5:0", shape=(None,), dtype=int64)), 'Sex': <tf.Tensor 'data_9:0' shape=(None,) dtype=int64>, 'Age': <tf.Tensor 'data:0' shape=(None,) dtype=float64>, 'SibSp': <tf.Tensor 'data_11:0' shape=(None,) dtype=int64>, 'Parch': <tf.Tensor 'data_6:0' shape=(None,) dtype=int64>, 'Ticket': <tf.Tensor 'data_12:0' shape=(None,) dtype=string>, 'Fare': <tf.Tensor 'data_3:0' shape=(None,) dtype=float64>, 'Cabin': <tf.Tensor 'data_1:0' shape=(None,) dtype=string>, 'Embarked': <tf.Tensor 'data_2:0' shape=(None,) dtype=string>, 'Title': <tf.Tensor 'data_14:0' shape=(None,) dtype=string>, 'Sex_Pclass': <tf.Tensor 'data_10:0' shape=(None,

[INFO 2026-09-06T06:33:24.708644235+00:00 kernel.cc:756] Start Yggdrasil model training
[INFO 2026-09-06T06:33:24.709499901+00:00 kernel.cc:757] Collect training examples
[INFO 2026-09-06T06:33:24.711841681+00:00 kernel.cc:388] Number of batches: 1
[INFO 2026-09-06T06:33:24.711871953+00:00 kernel.cc:389] Number of examples: 891
[INFO 2026-09-06T06:33:24.712755325+00:00 data_spec_inference.cc:303] 671 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Ticket_number (8 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:24.71300654+00:00 data_spec_inference.cc:303] 1 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Title (5 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:24.713205811+00:00 kernel.cc:774] Training dataset:
Number of records: 891
Number of columns: 6

Number of columns by type:
	CATEGORICAL: 5 (83.3333%)
	

Model trained in 0:00:00.221097
Compiling model...
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Model compiled.
Accuracy: 0.8260869383811951 Loss:0.8624987602233887


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [14]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=2, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=10,
    categorical_algorithm="RANDOM",
    max_depth=5,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.compile(metrics=["accuracy"])
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

Use 4 thread(s) for training
Use /tmp/tmpsf3dybr6 as temporary training directory
Reading training dataset...
Training tensor examples:
Features: {'PassengerId': <tf.Tensor 'data_7:0' shape=(None,) dtype=int64>, 'Pclass': <tf.Tensor 'data_8:0' shape=(None,) dtype=int64>, 'Name': tf.RaggedTensor(values=Tensor("data_4:0", shape=(None,), dtype=string), row_splits=Tensor("data_5:0", shape=(None,), dtype=int64)), 'Sex': <tf.Tensor 'data_9:0' shape=(None,) dtype=int64>, 'Age': <tf.Tensor 'data:0' shape=(None,) dtype=float64>, 'SibSp': <tf.Tensor 'data_11:0' shape=(None,) dtype=int64>, 'Parch': <tf.Tensor 'data_6:0' shape=(None,) dtype=int64>, 'Ticket': <tf.Tensor 'data_12:0' shape=(None,) dtype=string>, 'Fare': <tf.Tensor 'data_3:0' shape=(None,) dtype=float64>, 'Cabin': <tf.Tensor 'data_1:0' shape=(None,) dtype=string>, 'Embarked': <tf.Tensor 'data_2:0' shape=(None,) dtype=string>, 'Title': <tf.Tensor 'data_14:0' shape=(None,) dtype=string>, 'Sex_Pclass': <tf.Tensor 'data_10:0' shape=(None,

[INFO 2026-09-06T06:33:27.312438533+00:00 kernel.cc:756] Start Yggdrasil model training
[INFO 2026-09-06T06:33:27.31247192+00:00 kernel.cc:757] Collect training examples
[INFO 2026-09-06T06:33:27.312605202+00:00 kernel.cc:388] Number of batches: 1
[INFO 2026-09-06T06:33:27.312621919+00:00 kernel.cc:389] Number of examples: 891
[INFO 2026-09-06T06:33:27.313446476+00:00 data_spec_inference.cc:303] 671 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Ticket_number (8 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:27.313588196+00:00 data_spec_inference.cc:303] 1 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Title (5 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:27.313751272+00:00 kernel.cc:774] Training dataset:
Number of records: 891
Number of columns: 6

Number of columns by type:
	CATEGORICAL: 5 (83.3333%)
	

Model trained in 0:00:00.338107
Compiling model...
Model compiled.
Accuracy: 0.8260869383811951 Loss:0.856290876865387


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [15]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (5):
	Age
	Embarked
	Sex_Pclass
	Ticket_number
	Title

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.    "Sex_Pclass"  0.469417 ################
    2.           "Age"  0.384806 ##########
    3. "Ticket_number"  0.284567 ###
    4.         "Title"  0.277402 ##
    5.      "Embarked"  0.240010 

Variable Importance: NUM_AS_ROOT:
    1.    "Sex_Pclass" 76.000000 ################
    2. "Ticket_number" 30.000000 #####
    3.           "Age" 12.000000 #
    4.         "Title"  6.000000 
    5.      "Embarked"  3.000000 

Variable Importance: NUM_NODES:
    1.           "Age" 658.000000 ##############

In [16]:
model_with_importance = tfdf.keras.GradientBoostedTreesModel(
    verbose=2,
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True,
    compute_permutation_variable_importance=True,
    random_seed=1234,
)
model_with_importance.fit(train_ds)
model_with_importance.make_inspector().variable_importances()

Use 4 thread(s) for training
Use /tmp/tmp02rsczkr as temporary training directory
Reading training dataset...
Training tensor examples:
Features: {'PassengerId': <tf.Tensor 'data_7:0' shape=(None,) dtype=int64>, 'Pclass': <tf.Tensor 'data_8:0' shape=(None,) dtype=int64>, 'Name': tf.RaggedTensor(values=Tensor("data_4:0", shape=(None,), dtype=string), row_splits=Tensor("data_5:0", shape=(None,), dtype=int64)), 'Sex': <tf.Tensor 'data_9:0' shape=(None,) dtype=int64>, 'Age': <tf.Tensor 'data:0' shape=(None,) dtype=float64>, 'SibSp': <tf.Tensor 'data_11:0' shape=(None,) dtype=int64>, 'Parch': <tf.Tensor 'data_6:0' shape=(None,) dtype=int64>, 'Ticket': <tf.Tensor 'data_12:0' shape=(None,) dtype=string>, 'Fare': <tf.Tensor 'data_3:0' shape=(None,) dtype=float64>, 'Cabin': <tf.Tensor 'data_1:0' shape=(None,) dtype=string>, 'Embarked': <tf.Tensor 'data_2:0' shape=(None,) dtype=string>, 'Title': <tf.Tensor 'data_14:0' shape=(None,) dtype=string>, 'Sex_Pclass': <tf.Tensor 'data_10:0' shape=(None,

[INFO 2026-09-06T06:33:28.215638995+00:00 kernel.cc:756] Start Yggdrasil model training
[INFO 2026-09-06T06:33:28.215680939+00:00 kernel.cc:757] Collect training examples
[INFO 2026-09-06T06:33:28.215863649+00:00 kernel.cc:388] Number of batches: 1
[INFO 2026-09-06T06:33:28.215887021+00:00 kernel.cc:389] Number of examples: 891
[INFO 2026-09-06T06:33:28.216567814+00:00 data_spec_inference.cc:303] 671 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Ticket_number (8 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:28.21675353+00:00 data_spec_inference.cc:303] 1 item(s) have been pruned (i.e. they are considered out of dictionary) for the column Title (5 item(s) left) because min_value_count=5 and max_number_of_unique_values=2000
[INFO 2026-09-06T06:33:28.216974748+00:00 kernel.cc:774] Training dataset:
Number of records: 891
Number of columns: 6

Number of columns by type:
	CATEGORICAL: 5 (83.3333%)
	

Model trained in 0:00:00.185377
Compiling model...
Model compiled.


{'MEAN_DECREASE_IN_AP_2_VS_OTHERS': [("Sex_Pclass" (4; #2),
   0.2808977330164184),
  ("Title" (4; #4), 0.12049825638550782),
  ("Embarked" (4; #1), 0.02298440698124604),
  ("Ticket_number" (4; #3), 0.021366087742330286),
  ("Age" (1; #0), -0.018132492061678218)],
 'NUM_AS_ROOT': [("Sex_Pclass" (4; #2), 32.0),
  ("Title" (4; #4), 5.0),
  ("Ticket_number" (4; #3), 1.0)],
 'INV_MEAN_MIN_DEPTH': [("Sex_Pclass" (4; #2), 0.686364276999472),
  ("Age" (1; #0), 0.3394611720564485),
  ("Title" (4; #4), 0.2674193571283735),
  ("Ticket_number" (4; #3), 0.20867605370346096),
  ("Embarked" (4; #1), 0.19855359425360083)],
 'MEAN_DECREASE_IN_ACCURACY': [("Sex_Pclass" (4; #2), 0.15217387676239014),
  ("Title" (4; #4), 0.1086956262588501),
  ("Embarked" (4; #1), 0.03260868787765503),
  ("Age" (1; #0), 0.02173912525177002),
  ("Ticket_number" (4; #3), 0.02173912525177002)],
 'NUM_NODES': [("Age" (1; #0), 449.0),
  ("Sex_Pclass" (4; #2), 106.0),
  ("Title" (4; #4), 76.0),
  ("Embarked" (4; #1), 66.0),
  

# Make predictions

In [17]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,1
893,1
894,1
895,1
896,1
897,1
898,1
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [18]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpmom6qc3b as temporary training directory


[INFO 2026-09-06T06:35:25.623861543+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmom6qc3b/model/ with prefix bf9cf61ab9684def
[INFO 2026-09-06T06:35:25.636581233+00:00 decision_forest.cc:661] Model loaded with 24 root(s), 744 node(s), and 14 input feature(s).
[INFO 2026-09-06T06:35:25.636624203+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2026-09-06T06:35:25.636648202+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.8767123222351074 Loss:0.6209227442741394


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [19]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2026-09-06T06:35:26.676299474+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfy7je7o8/model/ with prefix d49137c753c64af6
[INFO 2026-09-06T06:35:26.680145042+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2026-09-06T06:35:27.461442843+00:00 kernel.cc:1214] Loading model from path /tmp/tmptlt_k5st/model/ with prefix 9055314bcbc643be
[INFO 2026-09-06T06:35:27.467646819+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2026-09-06T06:35:28.233015905+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7pj1jfna/model/ with prefix 6ca9d0c33663435f
[INFO 2026-09-06T06:35:28.238478779+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2026-09-06T06:35:29.00466202+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmtkg49sh/model/ with prefix 2e72826247c04eb6
[INFO 2026-09-06T06:35:29.010434481+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2026-09-06T06:35:29.867182244+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdsx4hxv9/model/ with prefix 363993f7435b47db
[INFO 2026-09-06T06:35:29.877607771+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2026-09-06T06:35:30.612198006+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkm4rkkzv/model/ with prefix 902b253fd5964df9
[INFO 2026-09-06T06:35:30.615588401+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2026-09-06T06:35:31.400348552+00:00 kernel.cc:1214] Loading model from path /tmp/tmphrkldlo9/model/ with prefix f57e3f6f23ec48bb
[INFO 2026-09-06T06:35:31.406803761+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2026-09-06T06:35:32.206072065+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqk_r0qmd/model/ with prefix da527888a36941a8
[INFO 2026-09-06T06:35:32.213116459+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2026-09-06T06:35:33.004146418+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmkjss_xm/model/ with prefix fee76180763944ab
[INFO 2026-09-06T06:35:33.010755841+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2026-09-06T06:35:34.281695424+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcgmfuawb/model/ with prefix 1ad4c7e8e1794590
[INFO 2026-09-06T06:35:34.29254972+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2026-09-06T06:35:35.060995434+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgt_ailhu/model/ with prefix 81635f4cfa1a4cfb
[INFO 2026-09-06T06:35:35.065768147+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2026-09-06T06:35:35.825299908+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuuidxk15/model/ with prefix 015b7a73e6c347b7
[INFO 2026-09-06T06:35:35.830666173+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:35:35.830703517+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2026-09-06T06:35:36.699552909+00:00 kernel.cc:1214] Loading model from path /tmp/tmppy54zmdp/model/ with prefix 8c2e521b130e45a6
[INFO 2026-09-06T06:35:36.709743595+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2026-09-06T06:35:37.615894498+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjyxf45yc/model/ with prefix bb49b71395714314
[INFO 2026-09-06T06:35:37.630213618+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2026-09-06T06:35:38.390354263+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnn14xpi7/model/ with prefix 945c384341474128
[INFO 2026-09-06T06:35:38.394695978+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2026-09-06T06:35:39.127904004+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzmg5nrle/model/ with prefix 0b85177bfa454410
[INFO 2026-09-06T06:35:39.130922285+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2026-09-06T06:35:39.909806589+00:00 kernel.cc:1214] Loading model from path /tmp/tmplj49zq3o/model/ with prefix 3a09e96615db40c7
[INFO 2026-09-06T06:35:39.915741686+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2026-09-06T06:35:40.698493297+00:00 kernel.cc:1214] Loading model from path /tmp/tmp44lk1wfy/model/ with prefix 4d7a3913263a40ae
[INFO 2026-09-06T06:35:40.704316992+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2026-09-06T06:35:41.47973405+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl00g3jkx/model/ with prefix eeefee5e7a784762
[INFO 2026-09-06T06:35:41.48519771+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2026-09-06T06:35:42.288118884+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdvogq6eq/model/ with prefix 29df00fbf2b44902
[INFO 2026-09-06T06:35:42.294139748+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2026-09-06T06:35:43.150935707+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnfzgo5fy/model/ with prefix 457ff736aa2b4340
[INFO 2026-09-06T06:35:43.161763044+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2026-09-06T06:35:43.920365084+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr72emslm/model/ with prefix a1f78af1b127470c
[INFO 2026-09-06T06:35:43.924206558+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2026-09-06T06:35:44.678841548+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7hb4rim_/model/ with prefix c2707059a6ff4f44
[INFO 2026-09-06T06:35:44.682644231+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2026-09-06T06:35:45.643153892+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx9m4vl9c/model/ with prefix d2be7111c0d244ec
[INFO 2026-09-06T06:35:45.659921411+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2026-09-06T06:35:46.469383932+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfnr8a6kw/model/ with prefix 55bc30cfcb984af9
[INFO 2026-09-06T06:35:46.475404974+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:35:46.475452843+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2026-09-06T06:35:47.414913891+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc2b8wiia/model/ with prefix ef2dd82156ad401f
[INFO 2026-09-06T06:35:47.431794604+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2026-09-06T06:35:48.236017348+00:00 kernel.cc:1214] Loading model from path /tmp/tmpse5jzs8k/model/ with prefix f6a4b473c8874bc8
[INFO 2026-09-06T06:35:48.243710825+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2026-09-06T06:35:49.021311891+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnktzphy4/model/ with prefix 2d65f5766ebe48ff
[INFO 2026-09-06T06:35:49.027244806+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2026-09-06T06:35:49.78692769+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4wqhljsy/model/ with prefix 765c769e4c854722
[INFO 2026-09-06T06:35:49.791856288+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2026-09-06T06:35:50.706639719+00:00 kernel.cc:1214] Loading model from path /tmp/tmppgob_rpa/model/ with prefix 2b2f89496eb9423d
[INFO 2026-09-06T06:35:50.720755119+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2026-09-06T06:35:51.600627298+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwf40eq0w/model/ with prefix fd90dceca5864a43
[INFO 2026-09-06T06:35:51.611494415+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2026-09-06T06:35:53.095900602+00:00 kernel.cc:1214] Loading model from path /tmp/tmppyg5chpl/model/ with prefix b5b77238ddc44f3d
[INFO 2026-09-06T06:35:53.115404163+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2026-09-06T06:35:53.964050932+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_v2a0xjb/model/ with prefix 15383fed26ea42d9
[INFO 2026-09-06T06:35:53.971428167+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2026-09-06T06:35:54.816867547+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmc0hyf6i/model/ with prefix 6bd12dc778924f56
[INFO 2026-09-06T06:35:54.826105896+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2026-09-06T06:35:55.621869625+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkhxfg8pz/model/ with prefix e0bb8512ca044a9b
[INFO 2026-09-06T06:35:55.626912876+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2026-09-06T06:35:56.459270414+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvyxe6u8_/model/ with prefix f8bfab98d6cc4543
[INFO 2026-09-06T06:35:56.465562277+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2026-09-06T06:35:57.291482275+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9q_2ndji/model/ with prefix faa5a6291e4f40bc
[INFO 2026-09-06T06:35:57.298303441+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:35:57.298344378+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2026-09-06T06:35:58.089777875+00:00 kernel.cc:1214] Loading model from path /tmp/tmplej3j7vz/model/ with prefix 4e77b5212e4c4c4d
[INFO 2026-09-06T06:35:58.095832003+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2026-09-06T06:35:58.906836141+00:00 kernel.cc:1214] Loading model from path /tmp/tmptvy1wlbo/model/ with prefix 7af311612484449a
[INFO 2026-09-06T06:35:58.913659289+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2026-09-06T06:35:59.853778954+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzhrcwk7s/model/ with prefix c9249524de964d22
[INFO 2026-09-06T06:35:59.870112081+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2026-09-06T06:36:00.633829227+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_0edjpki/model/ with prefix 60e40956cec34492
[INFO 2026-09-06T06:36:00.638265241+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2026-09-06T06:36:01.45844855+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3gmpdfk_/model/ with prefix dbcae7ceb62f46e8
[INFO 2026-09-06T06:36:01.466986102+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2026-09-06T06:36:02.232648549+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5uvf7kk5/model/ with prefix 4639b0653fbc4305
[INFO 2026-09-06T06:36:02.236812328+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2026-09-06T06:36:03.050358475+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo1u_61nr/model/ with prefix 019113e046e143b6
[INFO 2026-09-06T06:36:03.056740073+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2026-09-06T06:36:03.817759004+00:00 kernel.cc:1214] Loading model from path /tmp/tmp18jeu5bn/model/ with prefix d13a7e070b564750
[INFO 2026-09-06T06:36:03.820790923+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2026-09-06T06:36:04.593295787+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0s6526l6/model/ with prefix 6b69c2080d7344b2
[INFO 2026-09-06T06:36:04.597630679+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2026-09-06T06:36:05.507947361+00:00 kernel.cc:1214] Loading model from path /tmp/tmpseri1ctb/model/ with prefix d32315bba7744606
[INFO 2026-09-06T06:36:05.520473858+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2026-09-06T06:36:06.405712093+00:00 kernel.cc:1214] Loading model from path /tmp/tmpis4i62pi/model/ with prefix ceb1d153dd324a21
[INFO 2026-09-06T06:36:06.415178792+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2026-09-06T06:36:07.205302886+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8d1y4t95/model/ with prefix dd9f60da1ff34d5d
[INFO 2026-09-06T06:36:07.210032856+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2026-09-06T06:36:08.033542461+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4cl0chtx/model/ with prefix fa7a00214c994136
[INFO 2026-09-06T06:36:08.04049006+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:36:08.040532735+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2026-09-06T06:36:09.094955058+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9mz_6lt4/model/ with prefix eed0147ef29248a5
[INFO 2026-09-06T06:36:09.118280895+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2026-09-06T06:36:09.880258394+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm6xvubaj/model/ with prefix 387aa858fa534cc5
[INFO 2026-09-06T06:36:09.884800913+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2026-09-06T06:36:10.657012793+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwunrnsh9/model/ with prefix 3b2f72a631a24bab
[INFO 2026-09-06T06:36:10.661918122+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2026-09-06T06:36:11.512072802+00:00 kernel.cc:1214] Loading model from path /tmp/tmprcsdix9j/model/ with prefix 382a4dc33b504985
[INFO 2026-09-06T06:36:11.520356905+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2026-09-06T06:36:12.288346659+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq47hf41r/model/ with prefix f6bcc868368441ef
[INFO 2026-09-06T06:36:12.292106171+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2026-09-06T06:36:13.74144499+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzyvutnoj/model/ with prefix f2e9d9d8bfcb48d6
[INFO 2026-09-06T06:36:13.757493676+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2026-09-06T06:36:14.700765093+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjygnn3y3/model/ with prefix 7f1d76b93bd544dd
[INFO 2026-09-06T06:36:14.710825377+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2026-09-06T06:36:15.557824039+00:00 kernel.cc:1214] Loading model from path /tmp/tmp25hoodm2/model/ with prefix 7f943724037f4ee2
[INFO 2026-09-06T06:36:15.563626165+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2026-09-06T06:36:16.344945888+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7xyehqcc/model/ with prefix df0b957979c14ea4
[INFO 2026-09-06T06:36:16.348841281+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2026-09-06T06:36:17.281213504+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0uuwiop7/model/ with prefix 4780e6d3d28e4d5b
[INFO 2026-09-06T06:36:17.293042718+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2026-09-06T06:36:18.132970933+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt83ehrp7/model/ with prefix 525ea21cd233433c
[INFO 2026-09-06T06:36:18.140520478+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:36:18.140562192+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2026-09-06T06:36:18.94668715+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoy0sh8r4/model/ with prefix 6b76d3bc1c654164
[INFO 2026-09-06T06:36:18.951447159+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2026-09-06T06:36:19.765400864+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbzkr35d7/model/ with prefix 4f88e4767dd94a75
[INFO 2026-09-06T06:36:19.771834896+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2026-09-06T06:36:20.592952119+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsmlfnuyf/model/ with prefix a557c9aa3cd34a11
[INFO 2026-09-06T06:36:20.600313783+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2026-09-06T06:36:21.468667991+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4za24u3m/model/ with prefix b99ba5108a544af9
[INFO 2026-09-06T06:36:21.478728717+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2026-09-06T06:36:22.259613728+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnad60fth/model/ with prefix 92f0ee1bf742477a
[INFO 2026-09-06T06:36:22.263583612+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2026-09-06T06:36:23.151312183+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4qn9ry7x/model/ with prefix e1c0c30146bc4cc9
[INFO 2026-09-06T06:36:23.162756887+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2026-09-06T06:36:23.980658058+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj400o1tt/model/ with prefix 9f9febd5c2734902
[INFO 2026-09-06T06:36:23.987243196+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2026-09-06T06:36:24.820746778+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq9my_u8a/model/ with prefix 89cfefc4164047cc
[INFO 2026-09-06T06:36:24.827917973+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2026-09-06T06:36:25.628800656+00:00 kernel.cc:1214] Loading model from path /tmp/tmp007xwh7x/model/ with prefix f08f265917c54b26
[INFO 2026-09-06T06:36:25.634190226+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2026-09-06T06:36:26.452427734+00:00 kernel.cc:1214] Loading model from path /tmp/tmplwfgxs0q/model/ with prefix 5541df31c7b343a1
[INFO 2026-09-06T06:36:26.45707187+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2026-09-06T06:36:27.236668415+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9aspc9l7/model/ with prefix 2adfefabd22d4b4d
[INFO 2026-09-06T06:36:27.240452319+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2026-09-06T06:36:28.07030344+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9u1d_6g8/model/ with prefix 7119628eb6094248
[INFO 2026-09-06T06:36:28.078580142+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2026-09-06T06:36:28.894620179+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy01r11ou/model/ with prefix 96cab4a30a3646ca
[INFO 2026-09-06T06:36:28.90064391+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:36:28.900679242+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2026-09-06T06:36:29.754797342+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_uzb1rc0/model/ with prefix 8c6ca55d9afc4771
[INFO 2026-09-06T06:36:29.763854698+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2026-09-06T06:36:30.574758379+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcs2hpecu/model/ with prefix 75e5f51238a54006
[INFO 2026-09-06T06:36:30.580965585+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2026-09-06T06:36:31.358862359+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6f_canfh/model/ with prefix 3fc93d7437f44d95
[INFO 2026-09-06T06:36:31.362367151+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2026-09-06T06:36:32.164249283+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxb48mf82/model/ with prefix c267624a3629445d
[INFO 2026-09-06T06:36:32.169967351+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2026-09-06T06:36:32.97521954+00:00 kernel.cc:1214] Loading model from path /tmp/tmpasjury26/model/ with prefix 48777aca81c141d8
[INFO 2026-09-06T06:36:32.981263758+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2026-09-06T06:36:33.80644496+00:00 kernel.cc:1214] Loading model from path /tmp/tmpslwty9s8/model/ with prefix 010fbf5cc8dc4ad8
[INFO 2026-09-06T06:36:33.813595417+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2026-09-06T06:36:34.737398884+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkhxv7xcc/model/ with prefix f239c7e9b7584e4c
[INFO 2026-09-06T06:36:34.74990314+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2026-09-06T06:36:35.613597743+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdq1owlyr/model/ with prefix 404b1a435fd44b00
[INFO 2026-09-06T06:36:35.622181702+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2026-09-06T06:36:36.446524242+00:00 kernel.cc:1214] Loading model from path /tmp/tmpex4c2dh5/model/ with prefix 0e44533bc3ef4666
[INFO 2026-09-06T06:36:36.451034341+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2026-09-06T06:36:37.975510871+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1k74ynbu/model/ with prefix 1c3d577ee2e048f5
[INFO 2026-09-06T06:36:37.982954336+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2026-09-06T06:36:38.953856351+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb8hg3_u_/model/ with prefix d988a4c303c64bdd
[INFO 2026-09-06T06:36:38.965725673+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:36:38.965777943+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2026-09-06T06:36:39.78159328+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg3nppjwl/model/ with prefix 497531761634410e
[INFO 2026-09-06T06:36:39.785270812+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2026-09-06T06:36:40.657578019+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf5utb6ho/model/ with prefix 14f2b4792d744d9b
[INFO 2026-09-06T06:36:40.665483039+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2026-09-06T06:36:41.655919136+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbb32_rqe/model/ with prefix 247654004024499f
[INFO 2026-09-06T06:36:41.672910801+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2026-09-06T06:36:42.520427754+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw3r4bld0/model/ with prefix efe52751881d4ed7
[INFO 2026-09-06T06:36:42.52544355+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2026-09-06T06:36:43.343375326+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkjcn6scp/model/ with prefix 74619e32d8cf4a73
[INFO 2026-09-06T06:36:43.346784136+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2026-09-06T06:36:44.203227569+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxa9d82k3/model/ with prefix d61e2686a3be469a
[INFO 2026-09-06T06:36:44.209878493+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2026-09-06T06:36:45.069912097+00:00 kernel.cc:1214] Loading model from path /tmp/tmp26ors0dr/model/ with prefix dae43f21e33d4136
[INFO 2026-09-06T06:36:45.076551119+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2026-09-06T06:36:46.101006469+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl859xlm3/model/ with prefix d360ecb963cf4ce9
[INFO 2026-09-06T06:36:46.120263361+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2026-09-06T06:36:46.930465743+00:00 kernel.cc:1214] Loading model from path /tmp/tmphcpdu2hi/model/ with prefix 4b36eacc5c72456c
[INFO 2026-09-06T06:36:46.934432181+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2026-09-06T06:36:47.744721573+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo9eeojid/model/ with prefix 7682764ea5154936
[INFO 2026-09-06T06:36:47.749947886+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2026-09-06T06:36:48.682919704+00:00 kernel.cc:1214] Loading model from path /tmp/tmps7a6drkm/model/ with prefix b478d56575e14e43
[INFO 2026-09-06T06:36:48.696929693+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2026-09-06T06:36:49.521175767+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9h893gij/model/ with prefix f700ee07eb8346b3
[INFO 2026-09-06T06:36:49.527868645+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-09-06T06:36:49.527907095+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2026-09-06T06:36:50.37717482+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3kb49rpx/model/ with prefix 64ed855005ca4fd1
[INFO 2026-09-06T06:36:50.385010137+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2026-09-06T06:36:51.182642705+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz8a5tvqo/model/ with prefix d4f39a560cd94df9
[INFO 2026-09-06T06:36:51.187194489+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2026-09-06T06:36:52.096324413+00:00 kernel.cc:1214] Loading model from path /tmp/tmpje5okqbz/model/ with prefix 062401ee2ac2420d
[INFO 2026-09-06T06:36:52.106334087+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv
